scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java

```diff
@@ -201,8 +231,44 @@ public Long getLastModified() {
 
   public void refresh() {
     repositoryLocationResolver.refresh();
-    byNamespaceAndName.clear();
+    withWriteLockedByNamespaceMap(byNamespaceAndName::clear);
     byId.clear();
     init();
   }
+
+  private void withWriteLockedByNamespaceMap(Runnable runnable) {
+    Lock lock = byNamespaceLock.writeLock();
+    lock.lock();
+    try {
+      runnable.run();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private <T> T withReadLockedByNamespaceMap(Supplier<T> runnable) {
+    Lock lock = byNamespaceLock.readLock();
+    lock.lock();
+    try {
+      return runnable.get();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private static class RepositoryContainer {
```

Comment:
I don't get the need of the container. Why do we need it? Why don't we just use the repository without this wrapper?

Reply:
It's gone...